src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -360,20 +403,320 @@ duplicateIncomingRun ::
+  -> Ref (Run m h)
+  -> m (IncomingRun m h)
+newIncomingSingleRun tr ln r = do
+    r' <- dupRef r
+    traceWith tr $ AtLevel ln $
+      TraceNewMergeSingleRun (Run.size r') (Run.runFsPathsNumber r')
+    return (Single r')
+
+{-# SPECIALISE newIncomingCompletedMergingRun ::
+     Tracer IO (AtLevel MergeTrace)
+  -> TableConfig
+  -> LevelNo
+  -> MergePolicyForLevel
+  -> NumRuns
+  -> MergeDebt
+  -> Ref (Run IO h)
+  -> IO (IncomingRun IO h) #-}
+newIncomingCompletedMergingRun ::
+    (MonadMask m, MonadMVar m, MonadSTM m, MonadST m)
+  => Tracer m (AtLevel MergeTrace)
+  -> TableConfig
+  -> LevelNo
+  -> MergePolicyForLevel
+  -> NumRuns
+  -> MergeDebt
+  -> Ref (Run m h)
+  -> m (IncomingRun m h)
+newIncomingCompletedMergingRun tr conf ln mergePolicy nr mergeDebt r = do
+    traceWith tr $ AtLevel ln $
+      TraceNewMergeCompletedRun (Run.size r) (Run.runFsPathsNumber r)
+    mr <- MR.newCompleted nr mergeDebt r
+    let nominalDebt    = nominalDebtForLevel conf ln
+        nominalCredits = nominalDebtAsCredits nominalDebt
+    nominalCreditsVar <- newPrimVar nominalCredits
+    return (Merging mergePolicy nominalDebt nominalCreditsVar mr)
+
+{-# SPECIALISE newIncomingMergingRun ::
+     Tracer IO (AtLevel MergeTrace)
+  -> HasFS IO h
+  -> HasBlockIO IO h
+  -> ActiveDir
+  -> UniqCounter IO
+  -> TableConfig
+  -> ResolveSerialisedValue
+  -> MergePolicyForLevel
+  -> MR.LevelMergeType
+  -> LevelNo
+  -> V.Vector (Ref (Run IO h))
+  -> IO (IncomingRun IO h) #-}
+newIncomingMergingRun ::
+     (MonadMask m, MonadMVar m, MonadSTM m, MonadST m)
+  => Tracer m (AtLevel MergeTrace)
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> ActiveDir
+  -> UniqCounter m
+  -> TableConfig
+  -> ResolveSerialisedValue
+  -> MergePolicyForLevel
+  -> MR.LevelMergeType
+  -> LevelNo
+  -> V.Vector (Ref (Run m h))
+  -> m (IncomingRun m h)
+newIncomingMergingRun tr hfs hbio activeDir uc
+                      conf@TableConfig {
+                             confDiskCachePolicy,
+                             confFencePointerIndex
+                           }
+                      resolve mergePolicy mergeType ln rs = do
+    !rn <- uniqueToRunNumber <$> incrUniqCounter uc
+    let !caching   = diskCachePolicyForLevel confDiskCachePolicy ln
+        !alloc     = bloomFilterAllocForLevel conf ln
+        !indexType = indexTypeForRun confFencePointerIndex
+        !runPaths  = Paths.runPath activeDir rn
+    traceWith tr $ AtLevel ln $
+      TraceNewMerge (V.map Run.size rs) (runNumber runPaths)
+                    caching alloc mergePolicy mergeType
+    mr <- MR.new hfs hbio resolve caching
+                 alloc indexType mergeType
+                 runPaths rs
+    let nominalDebt    = nominalDebtForLevel conf ln
+        nominalCredits = NominalCredits 0
+    nominalCreditsVar <- newPrimVar nominalCredits
+    assert (MR.totalMergeDebt mr <= maxMergeDebt conf mergePolicy ln) $
+      return (Merging mergePolicy nominalDebt nominalCreditsVar mr)
```

Comment:
If the assert before the return fails, then this function will leak `mr`

Reply:
I don't think we need to worry about resource leaks due to assertion violations.